### ISSUE TITLE
Missing reportActions! 😱 

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#14424]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

HOLD on https://github.com/Expensify/App/pull/13118

## Action Performed:
0. On dev, do yourself a favor and change the pagination size from 50 to 5 [here](https://github.com/Expensify/Web-Expensify/blob/acef1b79a753fddfdb758159311e02f7869b9ca1/lib/ReportAPI.php#L4235). The bug is still valid with a larger page size, just easier to reproduce with a smaller one.

    ```diff
    + $indexedHistory = ReportAPI::getIndexedHistory($authToken, $reportID, -1, '', 5, Report::ACTION_ORDER_DESC, $shouldKeyReportActionsByID);
    - $indexedHistory = ReportAPI::getIndexedHistory($authToken, $reportID, -1, '', Report::ACTIONS_PAGINATION_LIMIT, Report::ACTION_ORDER_DESC, $shouldKeyReportActionsByID);
    ```
1. Sign in on two separate accounts, A and B.
1. On account A, send 5 messages to account B: `1`, `2`, `3`, `4`, `5`

    <img width="890" alt="Screenshot 2023-01-19 at 2 38 48 PM" src="https://user-images.githubusercontent.com/47436092/213579313-df160a16-26c9-4d48-80d4-e5c312519961.png">

1. As user A, go offline but keep the app open.
1. As user B, send 15 responses – 6 through 20

    <img width="1792" alt="Screenshot 2023-01-19 at 2 39 48 PM" src="https://user-images.githubusercontent.com/47436092/213579473-b8e0d3ad-aefa-45db-895c-7bc54b616789.png">

1. As user A, come back online.

## Expected Result:
- User A should see chats 1-20, without any missing, OR a more robust solution that depends on bidirectional pagination would be to:
- We display a skeleton UI between chats 6 and 16 to signify a gap where reportActions are missing. In production w/ our page size of 50, you probably would never see both sides of the "gap" on the screen at the same time. Scrolling into the the gap from above or below should trigger an API request to fill in the missing reportActions (or at least, one page of missing reportActions).

## Actual Result:
User A is missing chats 6-15:

<img width="1792" alt="Screenshot 2023-01-19 at 2 40 07 PM" src="https://user-images.githubusercontent.com/47436092/213579674-eae3ee15-fc73-4fac-8c8b-4764e8768299.png">

## Workaround:
None. Once you are in this state there's not really any way to get the missing chats.

## Platforms:
Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Issue reported by:** @roryabraham 
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1674106258093279?thread_ts=1674077185.040839&cid=C01GTK53T8Q

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013217438ced0fdf01</li>
        <li>Upwork Job ID: 1616207735730991104</li>
        <li>Last Price Increase: 2023-01-19</li>
    </ul>
</details>

**Comment by marcaaron** [comment]
naive solution that comes to mind is...

1. Send the most recent "non optimistic" (important) `reportActionID` the client currently has for the report they are opening
2. If they have none - load the most recent up to the limit
3. If they have one - then load the most recent from whatever action they do have

**Comment by roryabraham** [comment]
Yeah, the way I see it there are two high-level options:

1. Load all the reportActions between the most recent non-optimistic one you have and the canonical newest reportAction on the report.
    - Pro: simple
    - Con: doesn't scale well because this could be hundreds or thousands of reportActions if you're coming back to a very active room after a long time.
1. Display a skeleton UI in the gap where reportActions are missing. Scrolling into the the gap from above or below should trigger an API request to fill in the missing reportActions (or at least, one page of missing reportActions).
    - Pro: More robust, you can jump way back in chat history (like you might by clicking on a link to an old comment), and only load one page in that point of the history, then scroll up or down as you please.
    - Con: It's kind of hand-wavy and more thought is required to figure out how it would actually work.

**Comment by roryabraham** [comment]
Note: A different manifestation of this problem is likely to occur with comment linking:

1. You've got the newest 50 chats on a report loaded.
1. Then you click a link to a very old comment in the report. There's too much data between where you are now and that old comment to load it all at once, so you load just the page that contains that old comment.
1. Now you've got a gap.

This is kind of pushing me to think that while we could do the simple solution for now, if we need to to unblock something, but long-term the 2nd solution in my comment above will be needed.

**Comment by marcaaron** [comment]
>this could be hundreds or thousands of reportActions if you're coming back to a very active room after a long time

I kinda feel like if we are going to send this many actions back to a user we would just **not do that** and instead do something like... 

1. Wipe out their existing report actions (being careful to save any pending ones)
2. Send back the newest

In other words just ditch whatever old stuff they have so that there are no "gaps"

**Comment by marcaaron** [comment]
>Display a skeleton UI in the gap where reportActions are missing. Scrolling into the the gap from above or below should trigger an API request to fill in the missing reportActions (or at least, one page of missing reportActions)

Interesting, but the client won't know about things it doesn't know about. It doesn't know which comments belong where in the "canonical order" client just sees 5 random actions sorted by the date they were created and then another 5 actions that are more recent.

**Comment by marcaaron** [comment]
**Solves above problem, but with a bad UX:** on scroll -> call the API to see if there are gaps -> plop missing comments into the chat 🚢 

Sounds crazy at first, but might be a different way to think about this problem... 
Do we have a problem where gaps are possible?
Or are gaps and changes (e.g. edits, deletes) fine and we just need a system that detects and backfills inconsistencies...

**Comment by roryabraham** [comment]
Still pretty important, I just haven't had the will do tackle this problem yet 😓

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days.  eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by roryabraham** [comment]
This doesn't need to be on HOLD any more

**Comment by flodnv** [comment]
Remind me why this is happening? Maybe we can solve this with a `report.lastModified` [like mentioned here](https://github.com/Expensify/Expensify/issues/263266#issuecomment-1470403990)? When you come back online and call OpenReport, you pass to the API your lastModified, and the API sends you all updates since then?

**Comment by roryabraham** [comment]
This is still a thing.

**Comment by roryabraham** [comment]
I feel like we should put this on HOLD for comment linking, because the plan in that doc might actually fix this.

**Comment by roryabraham** [comment]
Let's confirm with @perunt

**Comment by roryabraham** [comment]
Posted in slack: https://expensify.slack.com/archives/C035J5C9FAP/p1686623759097729
